### ISSUE TITLE
Always add composer/installed.json before dumping autoload

### ISSUE THREAD

**Issue by frederikbosch** [humbug/box#256]
## Bug report

| Question    | Answer
| ------------| ---------------
| Box version | 3.0 beta 2 
| PHP version | 7.1.latest
| Platform with version | Ubuntu
| Github Repo | -

If someone uses a finder, e.g. the one below, the file `composer/installed.json` gets deleted. This file is however extremely important. If it is missing, the dumping of the autoloader generates unwanted results. Because then it leaves out all the non-dev dependencies from autoloading. This should be prevented.

```json
// this filters out composer/installed.json which is never desired
"finder": [
        {
            "name": "*.php",
            "exclude": ["Tests","test","Test","tests"],
            "in": "vendor"
        }
],
```

That is deleted after the dumping of the autoloader is not important, and that features can stay as is.

**Comment by frederikbosch** [comment]
Actaully, I have a lot more trouble getting everything to work correctly. It seems that my project will only work if I explicitly add the vendor folder.

```json
"directories": [
        "src", "vendor"
    ],
```

or

```json
"finder": [
        {
            "name": ["*"],
            "in": "vendor"
        }
    ],
```

**Comment by theofidry** [comment]
>the dumping of the autoloader generates unwanted results

As you noted the dumping autoloader occurs _before_ this file is removed, so it shouldn't be an issue.

> It seems that my project will only work if I explicitly add the vendor folder.

I cant' say for sure what's going there, it highly depends of what your application is doing and sometimes it can simply be a dependency doing something very unconventional. Note however that:

- You can let Box pick the files by itself, so leaving out the `finder` config and just cherry-pick some files and directories if necessary
- You can tell Box [not to exclude the Composer specific files if need be](https://github.com/humbug/box/blob/master/doc/configuration.md#excluding-the-composer-files-exclude-composer-files)

Note however that this last link is missing from the README ToC (it is in the config doc) so that's definitely a miss in the doc there

**Comment by frederikbosch** [comment]
@theofidry It is different. In order to dump the autoloader `installed.json` is required. See the steps below for a `box comile` execution.

1. In the end buffering call, the [dump of the autoloader is called](https://github.com/humbug/box/blob/master/src/Box.php#L132).
2. In the `ComposerOrchestrator` [a `Composer` instance is created](https://github.com/humbug/box/blob/master/src/Composer/ComposerOrchestrator.php#L42) through the `Composer\Factory`.
3. Through the `Composer` one [creates a autoload generator](https://github.com/humbug/box/blob/master/src/Composer/ComposerOrchestrator.php#L49).
4. Then you [dump the autoloader](https://github.com/humbug/box/blob/master/src/Composer/ComposerOrchestrator.php#L53).
5. The dump method requires to pass an `InstalledRepositoryInterface`. This `InstalledRepositoryInterface` contains a list of installed packages that should be used to create the correct dump of the autoloader. This was also [created through the `Composer` instance](https://github.com/humbug/box/blob/master/src/Composer/ComposerOrchestrator.php#L45).
6. The call to create an `InstalledRepositoryInterface` starts [in the `Composer\Factory`](https://github.com/composer/composer/blob/master/src/Composer/Factory.php#L339) and is [initialized in a sub-method of the `Composer\Factory`](https://github.com/composer/composer/blob/master/src/Composer/Factory.php#L429). As you can see a `Composer\Repository\InstalledFilesystemRepository` instance is created and the file `$vendorDir.'/composer/installed.json'` is passed as an argument.

This means, if `installed.json` is not added to the temporary directory, the autoloader will only dump the autoload for the root package and not include the autoload for the dependencies.

Hence, the file `composer/installed.json` should always be added to phar/temp dir before doing the autoload dump. It can be removed without a problem afterwards.

**Comment by theofidry** [comment]
>Hence, the file composer/installed.json should always be added to phar/temp dir before doing the autoload dump. It can be removed without a problem afterwards.

It is by default, it is not if you are specifying which files to pick with `finder`. In your case above, you filter `installed.json` because of the pattern `*.php`.

Maybe we could _always_ try to include it, but at the same I think the `finder` directive is for this but I'm not sure it's the way to go, instead I feel like if you need it and want the autoload to be dumped you should leave to Box how to find the files, i.e. omit the `finder` setting

**Comment by frederikbosch** [comment]
@theofidry Yes, the solution could also be documentation instead of code. 

However, my experience was that I also had to add the vendor folder to the `directories` setting in order to have my package compiled. So I don't know, but there seems to be more that is going on.

**Comment by theofidry** [comment]
>However, my experience was that I also had to add the vendor folder to the directories setting in order to have my package compiled. So I don't know, but there seems to be more that is going on.

That sounds weird: Box should be able to pick up the necessary files from `vendor` (ignoring the dev deps), provided everything is correctly registered in the `composer.lock` file. What kind of error are you getting if you don't specify `vendor`?

**Comment by theofidry** [comment]
Did you find anything regarding the second part of the issue?

**Comment by theofidry** [comment]
Closed by #260. Please open another issue for the second problem you found when you get more details!
